### ISSUE TITLE
Performance issues with Vite plugin

### ISSUE THREAD

**Issue by wenchonglee** [vanilla-extract-css/vanilla-extract#1211]
### Describe the bug

Hi,

I originally raised [an issue](https://github.com/mantinedev/mantine/issues/5184) in Mantine, but I don't think we can solve it from that end.

**TL;DR - it takes 30s to start the Vite dev server**
In the repo, it has 100 `.css.ts` files that imports from a common theme file. This theme file imports from `@mantine/core` and `@mantine/vanilla-extract`, but I don't think it's an issue with mantine.

I managed to reduce the start up time on my machine from 30s to 10s if I skip watching files in `node_modules`. I'm currently experimenting this with a pnpm patch as seen in this [commit](https://github.com/wenchonglee/ve-mantine-benchmark/commit/f73781ce980fbf9f3b8482ee0a6292e5394d8298#diff-cc160cdb1c61d54bfc5570f07e57484e9c36ca429561639124301328faf3f92a), but I'm not 100% sure it is sound. (10s is also not good enough for such a minimal repo)

I'm not very familiar with the internals of plugins and build tools, so I'd appreciate some insight into this and if it can/should be improved on Vanilla Extract's end.  

### Reproduction

https://github.com/wenchonglee/ve-mantine-benchmark

### System Info

```shell
System:
    OS: Windows 11 10.0.22621
    CPU: (12) x64 AMD Ryzen 5 3600 6-Core Processor
    Memory: 22.55 GB / 31.95 GB
  Binaries:
    Node: 20.9.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - C:\Program Files\nodejs\yarn.CMD
    npm: 10.1.0 - C:\Program Files\nodejs\npm.CMD
    pnpm: 8.10.2 - C:\Program Files\nodejs\pnpm.CMD
  Browsers:
    Edge: Chromium (118.0.2088.76)
    Internet Explorer: 11.0.22621.1
  npmPackages:
    @vanilla-extract/css: 1.13.0 => 1.13.0
    @vanilla-extract/dynamic: 2.0.3 => 2.0.3
    @vanilla-extract/integration: 6.2.2 => 6.2.2
    @vanilla-extract/recipes: 0.5.0 => 0.5.0
    @vanilla-extract/vite-plugin: 3.9.0 => 3.9.0
    vite: 4.5.0 => 4.5.0
```

### Used Package Manager

pnpm

### Validations

- [X] Check that there isn't [already an issue](https://github.com/vanilla-extract-css/vanilla-extract/issues) that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by mihkeleidast** [comment]
One thing I've come across in the past is that if a library exports vanilla-extract stuff (themes, utilities) from the same barrel index file with all other components etc, this will slow things down a lot. The solution for us was to export vanilla-extract things from a separate entrypoint.

Seems mantine is doing something like that, though I'm not familiar with that library, so can't say for sure.

**Comment by askoufis** [comment]
This is likely due to `@mantine/core`'s large barrel file export which exports theming and styling APIs as well as its entire component library.

This bottlenecks VE's style loading as it has to wait for esbuild (used by VE to compile styles) to build all of `@mantine/core` before loading any of the `.css.ts` files, and hence any of the components.

I'm fairly confident this is the issue as in a [CPU profile](https://github.com/vanilla-extract-css/vanilla-extract/files/13276378/ve-1211-repro.cpuprofile)
 I took of your reproduction there is a significant amount of idle time (~80%). This is likely the vite process waiting for the esbuild process to return the built mantine code.

Further evidence to this is the fact that I can reduce the first page load time from ~9s to ~2s on my machine by either commenting out all unnecessary exports in `node_modules/@mantine/core/esm/index.mjs` (need to keep about 5 exports around for it to work) _or_ by adding `@mantine/core` to `esbuildOptions.externals` in the VE vite plugin config. I _think_ the latter option is a viable solution, though not sure if that could have any unintended side-effects.

Loading all of `@mantine/core` is an unfair price to pay for users that want to use VE with mantine. I think an ideal solution could involve more granular endpoints on the `@mantine/core` package.

E.g.:
- `@mantine/core/components` for interactive components (i.e. not providers)
- `@mantine/core/utils` for APIs like `rgba`, `lighten`, `darken` etc.
- `@mantine/core/theme` for APis like `createTheme`, `MantineProvider`, `DEFAULT_THEME`

This would provide nice encapsulation of related APIs, while still keeping the code in the same core package, _and_ allowing for more granular tree-shaking to speed up consumer's builds (and VE styling obviously).

Of course I'm not really in a position to drastically influence the API design of mantine. This is a relatively naive assessment of the API though as I haven't used the library, so perhaps the execution of this would actually be quite difficult.

**Comment by wenchonglee** [comment]
Thanks for the really detailed explanation @askoufis! 
I can confirm your suggestion to `esbuildOptions.externals` also reduced my load time to around ~3s. We also won't need my patch anymore since `@mantine/core` code wouldn't be in the output in the first place.

I'll direct your post to the issue in Mantine and see if they are open to these changes, though I can see this being difficult or low on their priority as you've said. 

> I think the latter option is a viable solution, though not sure if that could have any unintended side-effects.

I guess the only question left is this? It seems fine but I can't say I know enough to be sure either.
If the VE team recommends this moving forward then I guess we can close this issue :)

**Comment by askoufis** [comment]
> any updates for this issue? it's mantine problem?

Yeah pretty much. There are workarounds, but IMO mantine should split up its API a bit more.

**Comment by mrm007** [comment]
@wenchonglee I tested with a prerelease version of the Vite plugin using the new compiler and observed a much better performance. On my machine, the server starts almost instantly and the CSS from each file is extracted in 1-6 seconds compared to 7-14 seconds with the current plugin.

You can try it out for yourself by installing [a snapshot version](https://github.com/vanilla-extract-css/vanilla-extract/actions/runs/7339066779) - `@vanilla-extract/vite-plugin@0.0.0-babel-modern-20231227140935`.

You'll have to set `emitCssInSsr: "compiler"` (subject to change) and run with `DEBUG=vanilla-extract pnpm dev` in [wenchonglee/ve-mantine-benchmark](https://github.com/wenchonglee/ve-mantine-benchmark).

<details>
<summary>With emitCssInSsr: true</summary>

```
➜ DEBUG=vanilla-extract nr dev

> ve-benchmark@0.0.0 dev /___/ve-mantine-benchmark
> vite

  VITE v4.5.0  ready in 302 ms

  ➜  Local:   http://localhost:5174/
  ➜  Network: use --host to expose
  ➜  press h to show help
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/9/styles.css.ts: 6.952s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/3/styles.css.ts: 7.141s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/15/styles.css.ts: 7.792s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/98/styles.css.ts: 7.886s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/24/styles.css.ts: 8.103s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/18/styles.css.ts: 8.370s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/99/styles.css.ts: 8.581s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/7/styles.css.ts: 8.750s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/4/styles.css.ts: 8.937s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/51/styles.css.ts: 9.099s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/17/styles.css.ts: 9.211s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/13/styles.css.ts: 9.322s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/12/styles.css.ts: 9.524s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/39/styles.css.ts: 9.636s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/26/styles.css.ts: 9.748s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/31/styles.css.ts: 9.862s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/32/styles.css.ts: 9.928s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/19/styles.css.ts: 9.985s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/49/styles.css.ts: 10.041s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/27/styles.css.ts: 10.098s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/29/styles.css.ts: 10.157s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/28/styles.css.ts: 10.213s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/36/styles.css.ts: 10.269s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/97/styles.css.ts: 10.319s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/11/styles.css.ts: 10.386s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/88/styles.css.ts: 10.442s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/42/styles.css.ts: 10.509s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/22/styles.css.ts: 10.567s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/63/styles.css.ts: 10.619s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/23/styles.css.ts: 10.677s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/45/styles.css.ts: 10.731s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/34/styles.css.ts: 10.788s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/8/styles.css.ts: 10.846s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/0/styles.css.ts: 10.901s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/80/styles.css.ts: 10.951s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/52/styles.css.ts: 11.016s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/10/styles.css.ts: 11.091s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/38/styles.css.ts: 11.148s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/83/styles.css.ts: 11.203s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/2/styles.css.ts: 11.270s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/14/styles.css.ts: 11.326s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/41/styles.css.ts: 11.382s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/48/styles.css.ts: 11.441s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/61/styles.css.ts: 11.496s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/55/styles.css.ts: 11.555s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/47/styles.css.ts: 11.612s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/50/styles.css.ts: 11.666s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/84/styles.css.ts: 11.715s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/16/styles.css.ts: 11.779s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/37/styles.css.ts: 11.833s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/59/styles.css.ts: 11.896s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/86/styles.css.ts: 11.950s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/78/styles.css.ts: 12.008s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/73/styles.css.ts: 12.063s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/6/styles.css.ts: 12.134s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/35/styles.css.ts: 12.179s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/1/styles.css.ts: 12.244s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/62/styles.css.ts: 12.298s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/93/styles.css.ts: 12.351s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/68/styles.css.ts: 12.409s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/69/styles.css.ts: 12.469s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/54/styles.css.ts: 12.528s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/5/styles.css.ts: 12.593s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/21/styles.css.ts: 12.653s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/96/styles.css.ts: 12.700s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/76/styles.css.ts: 12.764s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/75/styles.css.ts: 12.821s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/67/styles.css.ts: 12.877s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/95/styles.css.ts: 12.931s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/44/styles.css.ts: 13.001s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/64/styles.css.ts: 13.059s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/43/styles.css.ts: 13.118s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/89/styles.css.ts: 13.168s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/40/styles.css.ts: 13.233s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/87/styles.css.ts: 13.283s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/91/styles.css.ts: 13.341s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/81/styles.css.ts: 13.401s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/65/styles.css.ts: 13.462s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/20/styles.css.ts: 13.523s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/94/styles.css.ts: 13.572s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/79/styles.css.ts: 13.633s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/82/styles.css.ts: 13.690s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/46/styles.css.ts: 13.764s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/74/styles.css.ts: 13.820s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/57/styles.css.ts: 13.879s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/72/styles.css.ts: 13.934s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/70/styles.css.ts: 13.991s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/90/styles.css.ts: 14.041s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/25/styles.css.ts: 14.105s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/53/styles.css.ts: 14.160s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/56/styles.css.ts: 14.215s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/58/styles.css.ts: 14.277s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/66/styles.css.ts: 14.333s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/85/styles.css.ts: 14.387s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/33/styles.css.ts: 14.450s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/71/styles.css.ts: 14.503s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/30/styles.css.ts: 14.555s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/77/styles.css.ts: 14.616s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/60/styles.css.ts: 14.676s
[current] /___/ve-mantine-benchmark/src/ve-mantine-theme/92/styles.css.ts: 14.726s
```

</details>

<details>
<summary>With emitCssInSsr: 'compiler'</summary>

```
➜ DEBUG=vanilla-extract nr dev

> ve-benchmark@0.0.0 dev /___/ve-mantine-benchmark
> vite

  VITE v4.5.0  ready in 286 ms

  ➜  Local:   http://localhost:5174/
  ➜  Network: use --host to expose
  ➜  press h to show help
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/1/styles.css.ts: 1.337s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/6/styles.css.ts: 1.395s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/0/styles.css.ts: 1.449s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/2/styles.css.ts: 1.503s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/5/styles.css.ts: 1.559s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/4/styles.css.ts: 1.614s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/8/styles.css.ts: 1.668s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/3/styles.css.ts: 1.722s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/7/styles.css.ts: 1.778s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/9/styles.css.ts: 1.833s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/13/styles.css.ts: 1.891s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/11/styles.css.ts: 1.945s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/14/styles.css.ts: 2.001s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/15/styles.css.ts: 2.053s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/10/styles.css.ts: 2.111s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/17/styles.css.ts: 2.165s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/19/styles.css.ts: 2.221s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/12/styles.css.ts: 2.273s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/20/styles.css.ts: 2.327s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/18/styles.css.ts: 2.380s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/21/styles.css.ts: 2.435s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/23/styles.css.ts: 2.488s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/22/styles.css.ts: 2.541s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/24/styles.css.ts: 2.595s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/16/styles.css.ts: 2.649s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/25/styles.css.ts: 2.702s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/28/styles.css.ts: 2.758s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/27/styles.css.ts: 2.820s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/32/styles.css.ts: 2.880s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/30/styles.css.ts: 2.936s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/26/styles.css.ts: 2.990s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/29/styles.css.ts: 3.043s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/34/styles.css.ts: 3.103s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/36/styles.css.ts: 3.166s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/35/styles.css.ts: 3.220s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/33/styles.css.ts: 3.273s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/37/styles.css.ts: 3.327s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/31/styles.css.ts: 3.381s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/39/styles.css.ts: 3.438s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/41/styles.css.ts: 3.498s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/43/styles.css.ts: 3.554s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/38/styles.css.ts: 3.610s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/45/styles.css.ts: 3.677s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/40/styles.css.ts: 3.733s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/47/styles.css.ts: 3.788s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/42/styles.css.ts: 3.842s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/46/styles.css.ts: 3.896s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/49/styles.css.ts: 3.949s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/44/styles.css.ts: 4.003s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/54/styles.css.ts: 4.057s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/48/styles.css.ts: 4.117s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/53/styles.css.ts: 4.177s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/50/styles.css.ts: 4.235s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/51/styles.css.ts: 4.297s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/57/styles.css.ts: 4.353s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/58/styles.css.ts: 4.409s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/55/styles.css.ts: 4.465s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/56/styles.css.ts: 4.522s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/61/styles.css.ts: 4.578s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/60/styles.css.ts: 4.638s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/63/styles.css.ts: 4.698s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/62/styles.css.ts: 4.753s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/52/styles.css.ts: 4.807s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/59/styles.css.ts: 4.860s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/68/styles.css.ts: 4.914s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/67/styles.css.ts: 4.973s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/71/styles.css.ts: 5.036s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/65/styles.css.ts: 5.096s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/66/styles.css.ts: 5.148s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/70/styles.css.ts: 5.202s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/74/styles.css.ts: 5.255s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/64/styles.css.ts: 5.308s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/73/styles.css.ts: 5.362s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/75/styles.css.ts: 5.415s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/69/styles.css.ts: 5.469s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/76/styles.css.ts: 5.523s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/78/styles.css.ts: 5.581s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/80/styles.css.ts: 5.635s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/77/styles.css.ts: 5.689s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/72/styles.css.ts: 5.743s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/82/styles.css.ts: 5.796s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/79/styles.css.ts: 5.850s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/83/styles.css.ts: 5.904s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/86/styles.css.ts: 5.962s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/88/styles.css.ts: 6.019s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/87/styles.css.ts: 6.075s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/89/styles.css.ts: 6.130s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/84/styles.css.ts: 6.185s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/81/styles.css.ts: 6.239s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/91/styles.css.ts: 6.292s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/94/styles.css.ts: 6.346s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/90/styles.css.ts: 6.400s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/92/styles.css.ts: 6.454s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/85/styles.css.ts: 6.508s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/95/styles.css.ts: 6.562s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/98/styles.css.ts: 6.618s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/96/styles.css.ts: 6.673s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/93/styles.css.ts: 6.728s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/99/styles.css.ts: 6.782s
[compiler] /___/ve-mantine-benchmark/src/ve-mantine-theme/97/styles.css.ts: 6.837s
```

</details>

**Comment by wenchonglee** [comment]
Thanks for the update @mrm007, unfortunately my team and I decided to stop pursuing Vanilla Extract a month or so ago. 😅 

Though, I'd still be happy to help experiment if you need it. I ran the repo with your snapshot (I think you meant to write `0.0.0-vite-plugin-new-20231228080614` ) and I'm happy to report it went down from ~30s to ~16s for me. If I were to add the patch I did previously on top of the snapshot, it goes down further to ~6seconds. I understand that this issue is exacerbated by the way Mantine is written, so I'm not sure if you'd want to add that into the plugin.

**Comment by mrm007** [comment]
No worries! That patch seems like a good idea, will investigate https://github.com/vanilla-extract-css/vanilla-extract/pull/1264#discussion_r1437570197.

What made you not pursue Vanilla Extract? And what did you end up going with, if I may ask?

**Comment by wenchonglee** [comment]
Great, I'll keep an eye on the PR to see if I can learn something from it 😄 

Hmm, I'll say it's a combination of 3 primary concerns from the team. I was personally fine with all 3, especially since VE gives us type safety, but we ultimately decided CSS modules was generally "safer". Options like Tailwind were off the table since we're using Mantine.
1. This performance issue, it was worrying because we would have about ~200 `css.ts` files (it's a fairly large project). While there was some form of workaround, it still didn't come close to CSS modules. 
2. The [child->parent](https://vanilla-extract.style/documentation/styling/#complex-selectors) selector rule makes sense, but we weren't sure we agree with it being enforced. A few other VE APIs also adds up to having to write more code. 
3. Maintenance, CSS modules felt timeless and we won't have to deal with updates or any changes in API.  

I must say I understand it's a tradeoff between the two, my team valued stability more, so we went with what we deemed safer. The 2nd point in particular is not a criticism to VE. Hope that helps!
